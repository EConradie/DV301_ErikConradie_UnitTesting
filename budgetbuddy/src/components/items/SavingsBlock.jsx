import React from 'react'

function SavingsBlock(props) {
    const {savings} = props
  return (
    <div data-testid="savings-card" className='card card-col p-2 row-savings'>
        <div className='card-row'>
            <span>
                <div aria-label='icon' className='avatar'>
                    {savings.icon}
                </div>
                <p aria-label='name'>{savings.name}</p>
            </span>

        </div>
       
        <p aria-label='savings' className='amount amount-save'> R {savings.saves.toFixed(2)}<span> / R {savings.salary.toFixed(2)}</span></p>
    </div>
  )
}

export default SavingsBlock