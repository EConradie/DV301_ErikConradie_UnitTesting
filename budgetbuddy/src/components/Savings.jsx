import React, { useState, useEffect } from "react";
import TaxBlock from "./items/TaxBlock";
import { percentageOptions } from "../utils";
import { Form } from "react-bootstrap";
import SavingsBlock from "./items/SavingsBlock";
import { calculateSavings } from "../components/functions/SavingsCalculation"


function Savings({ income, onTotalSavings }) {
  const [selectedPercentage, setSelectedPercentage] = useState(null);

  const handlePercentageChange = (event) => {
    setSelectedPercentage(Number(event.target.value));
  };

  const { savingsData, totalSavings } = calculateSavings(income, selectedPercentage);

  useEffect(() => {
    if (onTotalSavings) {
      onTotalSavings(totalSavings);
    }
  }, [totalSavings, onTotalSavings]);

  return (
    <div data-testid="savings-component">
      <div className="title-row">
        <h3>Savings Calculation</h3>
        <span>
          <p>% you want to save</p>
          <Form.Select
            id="percentage"
            name="percentage"
            aria-label="percentage"
            defaultValue="-"
            autoComplete="off"
            onChange={handlePercentageChange}
          >
            <option disabled>-</option>
            {percentageOptions.map((amount, index) => (
              <option key={index} value={amount}>
                {amount}%
              </option>
            ))}
          </Form.Select>
        </span>
      </div>

      {/* List */}
      <div className="scroll-row hide-scroll">
        {savingsData &&
          savingsData.map((item, index) => (
            <SavingsBlock key={index} savings={item} />
          ))}
      </div>
    </div>
  );
}

export default Savings;
