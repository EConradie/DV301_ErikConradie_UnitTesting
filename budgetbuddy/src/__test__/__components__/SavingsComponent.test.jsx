import { render, screen } from '@testing-library/react';
import Savings from '../../components/Savings';
import SavingsBlock from '../../components/items/SavingsBlock';

test("Test default state of Savings Component", () => {
  render(<Savings income={[]} />);

  var fullComponent = screen.getByTestId("savings-component");
  var title = screen.getByText("Savings Calculation");
  var input = screen.getByRole("input");

  expect(title).toBeInTheDocument();
  expect(input).toBeInTheDocument();
  expect(fullComponent).toMatchSnapshot();
});

test("Test if savings cards display when 2 savings are added", () => {
  var savings = [
    {
      name: "Savings 1",
      amount: 1000,
      icon: "ðŸ’°",
    },
    {
      name: "Savings 2",
      amount: 2000,
      icon: "ðŸ’°",
    }
  ];

  render(
    <div>
      {savings.map((saving, index) => (
        <SavingsBlock key={index} saving={saving} />
      ))}
    </div>
  );

  var iconDisplay = screen.getAllByLabelText("icon");
  var nameDisplay = screen.getAllByLabelText("name");
  var savingsDisplay = screen.getAllByLabelText("savings");
  var savingsCards = screen.getAllByTestId("savings-card"); 

  savingsCards.forEach((element, index) => {

    expect(element).toBeInTheDocument();
    expect(element).toHaveTextContent(`R ${savings[index].amount.toFixed(2)}`);
    expect(element).toHaveTextContent(savings[index].name);
    expect(element).toHaveTextContent(savings[index].icon);

  })

  expect(iconDisplay).toHaveLength(2);
  expect(nameDisplay).toHaveLength(2);
  expect(savingsDisplay).toHaveLength(2);    
});