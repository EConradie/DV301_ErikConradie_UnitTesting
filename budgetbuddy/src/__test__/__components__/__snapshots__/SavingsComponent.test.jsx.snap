// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test default state of SavingsBlock Component 1`] = `
<div
  class="card card-col p-2 row-savings"
  data-testid="savings-card"
>
  <div
    class="card-row"
  >
    <span>
      <div
        aria-label="icon"
        class="avatar"
      >
        ðŸ’°
      </div>
      <p
        aria-label="name"
      >
        Sample Savings
      </p>
    </span>
  </div>
  <p
    aria-label="savings"
    class="amount amount-save"
  >
     R 
    200.00
    <span>
       / R 
      1000.00
    </span>
  </p>
</div>
`;
